### ISSUE TITLE
lint for unused crates

### ISSUE THREAD

**Issue by brson** [rust-lang/rust-clippy#4341]
It's pretty common for crates listed in `Cargo.toml` to become unused over time. I don't know of a tool that can figure this out, but it definitely needs the help of rustc. Clippy is in a position to figure this out. rustc should be able to track whether any names from a crate explicitly loaded by rustc are actually used and report when none are.

cc @siddontang

**Comment by brson** [comment]
Note that this is different from https://github.com/rust-lang/rust-clippy/issues/3513, about remove `extern crate` statements, but might also properly account for `extern crate` statements from which no names are resolved.

**Comment by matthiaskrgr** [comment]
This might be quite hard to get right because some imported crates might only be used with some features enabled or on some platforms that do not compile for the host machine that runs `cargo clippy`... :thinking:

**Comment by brson** [comment]
@matthiaskrgr yeah it may not be possible to do perfectly, but if not I would see that as a cargo bug. Cargo can build crates conditionally based on features, but maybe not target platform (not sure). Ideally, cargo would not be building any crates that are unused for any particular configuration.

**Comment by brson** [comment]
fwiw I'm poking at a temporary solution to just add the appropriate "used / unused" logic to rustc and log it as `info!`, so it can be scripted together with cargo tree as a rough heuristic. Not sure how far I'll get there.

**Comment by brson** [comment]
Per @ehuss 

> @brson I'm not sure if it would help, but there is a new -Z binary-dep-depinfo flag which will record which dependencies were actually used in the dep-info file.  You can then try to correlate the rlibs with the dependencies in Cargo.toml. It may not be easy, but at least rustc does keep track of which crates were actually used.

**Comment by est31** [comment]
See also https://github.com/rust-lang/rust/issues/57274

**Comment by est31** [comment]
FYI, while I'd like to have it implemented in upstream tools as well, I have released an [independent tool](https://github.com/est31/cargo-udeps) that does such an unused crates lint.

**Comment by brson** [comment]
Thanks @est31 !

**Comment by roblabla** [comment]
This already exists in rustc, in the form of the unused_crate_dependencies lint: https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unused-crate-dependencies

This issue should probably be closed?

**Comment by est31** [comment]
@roblabla that lint is allow by default because it doesn't work well with how cargo offers crates to rustc. That is, each cargo *package* has multiple *crates*, and each crate can use different dependencies. One example is main.rs and lib.rs in the same package. Another one is crates used by dev dependencies, as a dev dependency can e.g. only be used by doctests. Giving good error messages in all these instances basically requires this to be handled in cargo. I had a PR open but it's blocked by a cargo mode that at least *compiles* everything that can use dev dependencies in one run, so that cargo can then turn on the unused dependencies check for dev dependencies too: https://github.com/rust-lang/cargo/pull/8437
